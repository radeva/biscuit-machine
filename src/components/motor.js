import React from 'react';
import PropTypes from 'prop-types';
import { ReactComponent as MotorSVG } from './../images/motor.svg';
import useInterval from 'react-useinterval';

export const MOTOR_TIMEOUT = 4000;

export default function Motor(props) {
  let containerCss = 'motor';

  if (props.isMachineMovementPaused) {
    containerCss += ' motor-on animation-paused';
  }
  if (props.isOn) {
    containerCss += ' motor-on';
  }

  let intervalOffset = 0,
    lastIntervalStartDate = new Date();

  const handlePulse = () => {
    if (props.isMachineMovementPaused) {
      if (intervalOffset === 0) {
        intervalOffset = new Date().getTime() - lastIntervalStartDate.getTime();
      }

      return;
    }

    if (intervalOffset > 0) {
      setTimeout(() => {
        lastIntervalStartDate = new Date();
        props.onSendPulse();
      }, intervalOffset);
      intervalOffset = 0;
    } else {
      lastIntervalStartDate = new Date();
      props.onSendPulse();
    }
  };

  useInterval(
    handlePulse,
    props.isOn || props.isMachineMovementPaused ? MOTOR_TIMEOUT : null,
  );
  return (
    <div data-testid="motor-container" className={containerCss}>
      {' '}
      <MotorSVG className="motor-svg" />{' '}
    </div>
  );
}

Motor.propTypes = {
  isMachineMovementPaused: PropTypes.bool,
  isOn: PropTypes.bool,
  onSendPulse: PropTypes.func.isRequired,
};

Motor.defaultProps = {
  isMachineMovementPaused: false,
  isOn: false,
};
