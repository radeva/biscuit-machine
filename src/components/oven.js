import React, { useState } from 'react';
import PropTypes from 'prop-types';
import useInterval from 'react-useinterval';
import { SWITCH_STATES } from './switch';
import { ReactComponent as BakedBiscuitSVG } from './../images/baked-biscuit.svg';

export const OVEN_MIN_TEMPERATURE = 220,
  OVEN_MAX_TEMPERATURE = 240,
  HEAT_OVEN_STEP = 2,
  HEAT_OVEN_INTERVAL_IN_MSECONDS = 100,
  WARMUP_COOLDOWN_OVEN_INTERVAL_IN_MSECONDS = 1000,
  INITIAL_TEMPERATURE = 40;

export default function Oven(props) {
  const [temperature, setTemperature] = useState(INITIAL_TEMPERATURE),
    [isOn, setIsOn] = useState(false),
    [isWarmingUp, setIsWarmingUp] = useState(false);

  const isOvenOn =
    props.switchState !== SWITCH_STATES.OFF || props.hasBiscuitOnConveyor;

  if ((!isOn && isOvenOn) || (isOn && !isOvenOn)) {
    setIsOn(isOvenOn);
  }

  const shouldTemperatureGoUp = isOn && temperature < OVEN_MIN_TEMPERATURE,
    shouldTemperatureGoDown =
      !isOn &&
      temperature > INITIAL_TEMPERATURE &&
      temperature <= OVEN_MIN_TEMPERATURE,
    shouldWarmUp =
      (isOn || props.hasBiscuitOnConveyor) &&
      (temperature === OVEN_MIN_TEMPERATURE ||
        (isWarmingUp &&
          temperature > OVEN_MIN_TEMPERATURE &&
          temperature < OVEN_MAX_TEMPERATURE)),
    shouldCoolDown =
      temperature === OVEN_MAX_TEMPERATURE ||
      ((!isWarmingUp || !isOn) &&
        temperature > OVEN_MIN_TEMPERATURE &&
        temperature < OVEN_MAX_TEMPERATURE);

  const calculateTemperatureDown = () => {
    if (temperature === OVEN_MIN_TEMPERATURE) {
      props.handleOvenReady(false);
    }

    setTemperature(temperature - HEAT_OVEN_STEP);
  };

  const calculateTemperatureUp = () => {
    const newTemperature = temperature + HEAT_OVEN_STEP;
    setTemperature(newTemperature);

    if (newTemperature === OVEN_MIN_TEMPERATURE) {
      props.handleOvenReady(true);
    }
  };

  const coolDown = () => {
    setIsWarmingUp(false);
    setTemperature(temperature - HEAT_OVEN_STEP);
  };

  const warmUp = () => {
    setIsWarmingUp(true);
    setTemperature(temperature + HEAT_OVEN_STEP);
  };

  useInterval(
    calculateTemperatureUp,
    shouldTemperatureGoUp ? HEAT_OVEN_INTERVAL_IN_MSECONDS : null,
  );
  useInterval(
    calculateTemperatureDown,
    shouldTemperatureGoDown ? HEAT_OVEN_INTERVAL_IN_MSECONDS : null,
  );
  useInterval(
    coolDown,
    shouldCoolDown ? WARMUP_COOLDOWN_OVEN_INTERVAL_IN_MSECONDS : null,
  );
  useInterval(
    warmUp,
    shouldWarmUp ? WARMUP_COOLDOWN_OVEN_INTERVAL_IN_MSECONDS : null,
  );

  let ovenLightClassName = 'oven-light';
  if (isOn) {
    ovenLightClassName += ' ' + ovenLightClassName + '-on';
  }

  let biscuitElement = '',
    animationPausedClass = '';
  if (props.switchState === SWITCH_STATES.PAUSE) {
    animationPausedClass = 'animation-paused';
  }
  if (props.hasBiscuitToBake) {
    biscuitElement = (
      <BakedBiscuitSVG
        data-testid="biscuit-in-oven"
        className={'baked-biscuit ' + animationPausedClass}
      />
    );
  }

  return (
    <div className="oven right">
      <label>GORENJE</label>
      <span className={ovenLightClassName} data-testid="oven-light">
        &#8226;
      </span>
      <br />
      <label data-testid="oven-temperature">
        {temperature}
        <span>&#8451;</span>
      </label>
      <div>{biscuitElement}</div>
    </div>
  );
}

Oven.propTypes = {
  switchState: PropTypes.oneOf(Object.values(SWITCH_STATES)).isRequired,
  hasBiscuitOnConveyor: PropTypes.bool,
  handleOvenReady: PropTypes.func.isRequired,
  hasBiscuitToBake: PropTypes.bool,
};

Oven.defaultProps = {
  hasBiscuitOnConveyor: false,
  hasBiscuitToBake: false,
};
