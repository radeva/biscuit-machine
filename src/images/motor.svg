<svg width="98" height="98" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 98 98">
 <!-- Created with Method Draw - http://github.com/duopixel/Method-Draw/ -->
 <g>
  <title>background</title>
  <rect fill="#fff" id="canvas_background" height="100" width="100" y="-1" x="-1"/>
  <g display="none" overflow="visible" y="0" x="0" height="100%" width="100%" id="canvasGrid">
   <rect fill="url(#gridpattern)" stroke-width="0" y="0" x="0" height="100%" width="100%"/>
  </g>
 </g>
 <g>
  <title>Layer 1</title>
  <ellipse ry="48" rx="48" id="svg_1" cy="48.75" cx="48.75" stroke-width="1.5" stroke="#000" fill="#56aaff"/>
  <ellipse ry="9.749932" rx="9.749932" id="svg_2" cy="48.749999" cx="48.750009" stroke-width="1.5" stroke="#000" fill="#ffffff"/>
  <line fill-opacity="0" stroke-dasharray="5,5" stroke-linecap="null" stroke-linejoin="null" id="svg_3" y2="38.88297" x2="49.000001" y1="1.22689" x1="49.000001" stroke-opacity="null" stroke-width="1.5" stroke="#000" fill="none"/>
 </g>
</svg>