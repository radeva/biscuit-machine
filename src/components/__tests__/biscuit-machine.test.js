import React from 'react';
import BiscuitMachine from '../biscuit-machine';
import { HEAT_OVEN_INTERVAL_IN_MSECONDS } from '../oven';
import { MOTOR_TIMEOUT } from '../motor';
import { render, cleanup, fireEvent, act } from '@testing-library/react';

afterEach(cleanup);

jest.useFakeTimers();

describe('Biscuit Machine', () => {

  it('should keep machine state as it is when paused', () => {
    const { getByText, container } = render(<BiscuitMachine />);

    const buttonOn = getByText('ON');
    fireEvent.click(buttonOn);

    act(() => {
      jest.advanceTimersByTime(HEAT_OVEN_INTERVAL_IN_MSECONDS * 10 * 9); // go up to 220 C
      jest.advanceTimersByTime(MOTOR_TIMEOUT * 2); // move to 110
    });

    expect(container).toMatchSnapshot();

    const buttonPause = getByText('PAUSE');
    fireEvent.click(buttonPause);

    act(() => {
      jest.advanceTimersByTime(MOTOR_TIMEOUT);
    });

    // check that state doesn't change
    expect(container).toMatchSnapshot();
  });
});
