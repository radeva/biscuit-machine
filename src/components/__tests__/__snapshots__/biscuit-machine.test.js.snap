// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Biscuit Machine should keep machine state as it is when paused 1`] = `
<div>
  <div
    class="container"
  >
    <h1>
      BISCUIT MACHINE
    </h1>
    <div>
      <div
        class="biscuit-container"
      />
      <div
        class="extruder left"
      >
        <svg>
          extruder.svg
        </svg>
      </div>
      <div
        class="stamper left"
      >
        <svg>
          stamper.svg
        </svg>
        <svg
          class="stamper-pusher stamper-pusher-on "
          data-testid="stamper-pusher"
        >
          stamper-pusher.svg
        </svg>
        
      </div>
      <div
        class="oven right"
      >
        <label>
          GORENJE
        </label>
        <span
          class="oven-light oven-light-on"
          data-testid="oven-light"
        >
          •
        </span>
        <br />
        <label
          data-testid="oven-temperature"
        >
          228
          <span>
            ℃
          </span>
        </label>
      </div>
    </div>
    <div
      class="conveyor"
    >
      <svg
        class="sc-AxjAm jhTbDC"
      >
        conveyor.svg
      </svg>
    </div>
    <div>
      <div
        class="left"
      >
        <div
          class="motor motor-on"
          data-testid="motor-container"
        >
          <svg
            class="motor-svg"
          >
            motor.svg
          </svg>
        </div>
      </div>
      <div
        class="right"
      >
        <div
          class="switch-container"
        >
          <button
            class="switch-button switch-button-active"
          >
            ON
          </button>
          <button
            class="switch-button"
          >
            PAUSE
          </button>
          <button
            class="switch-button"
          >
            OFF
          </button>
        </div>
      </div>
    </div>
    <div
      class="biscuits-container"
    >
      <h2>
        Bucket with Biscuits (
        0
        )
      </h2>
      <ul
        class="biscuits-list"
        data-testid="biscuits-list"
      />
    </div>
  </div>
</div>
`;

exports[`Biscuit Machine should keep machine state as it is when paused 2`] = `
<div>
  <div
    class="container"
  >
    <h1>
      BISCUIT MACHINE
    </h1>
    <div>
      <div
        class="biscuit-container"
      >
        <svg
          class="biscuit-dough animation-paused"
          data-testid="biscuit-dough-svg"
        >
          biscuit-dough.svg
        </svg>
        <svg
          class="raw-biscuit animation-paused"
          data-testid="raw-biscuit"
        >
          raw-biscuit.svg
        </svg>
      </div>
      <div
        class="extruder left"
      >
        <svg>
          extruder.svg
        </svg>
      </div>
      <div
        class="stamper left"
      >
        <svg>
          stamper.svg
        </svg>
        <svg
          class="stamper-pusher stamper-pusher-on animation-paused"
          data-testid="stamper-pusher"
        >
          stamper-pusher.svg
        </svg>
        
      </div>
      <div
        class="oven right"
      >
        <label>
          GORENJE
        </label>
        <span
          class="oven-light oven-light-on"
          data-testid="oven-light"
        >
          •
        </span>
        <br />
        <label
          data-testid="oven-temperature"
        >
          232
          <span>
            ℃
          </span>
        </label>
      </div>
    </div>
    <div
      class="conveyor"
    >
      <svg>
        conveyor.svg
      </svg>
    </div>
    <div>
      <div
        class="left"
      >
        <div
          class="motor motor-on animation-paused"
          data-testid="motor-container"
        >
          <svg
            class="motor-svg"
          >
            motor.svg
          </svg>
        </div>
      </div>
      <div
        class="right"
      >
        <div
          class="switch-container"
        >
          <button
            class="switch-button"
          >
            ON
          </button>
          <button
            class="switch-button switch-button-active"
            disabled=""
          >
            PAUSE
          </button>
          <button
            class="switch-button"
          >
            OFF
          </button>
        </div>
      </div>
    </div>
    <div
      class="biscuits-container"
    >
      <h2>
        Bucket with Biscuits (
        0
        )
      </h2>
      <ul
        class="biscuits-list"
        data-testid="biscuits-list"
      />
    </div>
  </div>
</div>
`;
